File: php/testcode/fileupload_008.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_upload_out_of_webroot
function fileupload008(BenchmarkRequest $req): BenchmarkResponse {
    $file = $req->file('f');
    if (!$file) {
        return BenchmarkResponse::badRequest('no file');
    }
    $ext = strtolower(pathinfo($file['name'], PATHINFO_EXTENSION));
    $allowed = ['jpg', 'png', 'gif', 'pdf'];
    if (!in_array($ext, $allowed, true)) {
        return BenchmarkResponse::badRequest('invalid extension');
    }
    $safeName = bin2hex(random_bytes(16)) . '.' . $ext;
    $dest = '/var/uploads/' . $safeName; // vuln-code-snippet safe-line php_upload_out_of_webroot
    move_uploaded_file($file['tmp_name'], $dest);
    return BenchmarkResponse::ok("stored as $safeName");
}
// vuln-code-snippet end php_upload_out_of_webroot

File: ruby/testcode/fileupload_008.rb
require_relative 'shared'

PERMITTED_EXT = %w[jpg png gif pdf].freeze
STORAGE_ROOT = '/var/private/uploads'.freeze

# vuln-code-snippet start ruby_fileupload_outside_webroot
def upload_safe_storage(req)
  upload = req.file('file')
  ext = File.extname(upload[:filename]).delete('.').downcase
  return BenchmarkResponse.bad_request('not allowed') unless PERMITTED_EXT.include?(ext)
  safe_name = SecureRandom.uuid + '.' + ext
  File.write(File.join(STORAGE_ROOT, safe_name), upload[:data]) # vuln-code-snippet safe-line ruby_fileupload_outside_webroot
  BenchmarkResponse.ok(safe_name)
end
# vuln-code-snippet end ruby_fileupload_outside_webroot
